File: typescript/typescript-sorting/bubble.ts
let bubbleSwaps: number = 0;
let bubbleChecks: number = 0;

console.log("\nBubble Sort")

const bubbleSort = (array: number[]): number[] => {

    const newArray: number[] = structuredClone(array);
    let j: number = newArray.length;

    while(j > 0) {
        for(let i: number = 1; i < j; i++ ) {
            bubbleChecks++
            if(newArray[i-1] > newArray[i]) {
                const temp = newArray[i-1];
                newArray[i-1] = newArray[i];
                newArray[i] = temp;
                bubbleSwaps++
            }
        }
    j--;
    }

    return newArray;
}

export { bubbleSort, bubbleSwaps, bubbleChecks };

File: lua/lua-sorting/bubble.lua
local deepClone = require "functions/deepClone"

local sort = {}

sort.checks = 0
sort.swaps = 0

function sort.bubble(array)
    print("\nBubble Sort")

    local newArray = deepClone(array)
    local j = #newArray

    while j > 0 do
        for i = 2, j do
            sort.checks = sort.checks + 1
            if(newArray[i-1] > newArray[i]) then
                local temp = newArray[i]
                newArray[i] = newArray[i-1]
                newArray[i-1] = temp
                sort.swaps = sort.swaps + 1
            end
        end
        j = j - 1
    end
    return newArray
end

return sort
